### PR TITLE
Promises instead callbacks. Now server not crashed if send bad token.

### PR DESCRIPTION
**Pull request by OsirisAnubiz**
1. All callback into `storage.js` to promise (async/await).
1.1. Now function `safePath` takes a promise as the first argument and return promise, instead callback contract.
`const safePath = (fn) => (token, ...args) => {`
`const safePath = (promise) => (token, ...args) => {`
1.2. All methods in Storage async. This method await safePath promises.
`async get(key) {`
`data = await readSession(key);`
`async save(key) {`
`await writeSession(key, data);`
`async delete(key) {`
`await deleteSession(key);`
2. Storage return `undefined` if not found file session.
```
try {
  data = await readSession(key);
} catch (err) {
  return undefined;
}
```
This makes Storage contract look more like Map contract(if not found return `undefined`).
3. Session.restore method
```
static restore(client) {
  const { cookie } = client;
  if (!cookie) return;
  const sessionToken = cookie.token;
  if (sessionToken) {
    return new Promise((resolve, reject) => {
      storage.get(sessionToken, (err, session) => {
        if (err) reject(new Error('No session'));
        Object.setPrototypeOf(session, Session.prototype);
        client.token = sessionToken;
        client.session = session;
        resolve(session);
      });
    });
  }
}
```
```
static async restore(client) {
  const { cookie } = client;
  if (!cookie) return;
  const sessionToken = cookie.token;
  if (!sessionToken) return;
  const session = await storage.get(sessionToken);
  if (!session) return;
  Object.setPrototypeOf(session, Session.prototype);
  client.token = sessionToken;
  client.session = session;
  return session;
}
```
Now if you pass a bad token, the server will not crash.

### CODE REVIEW

File: JavaScript/storage.js [hunk 1]
```diff
@@ -1,63 +1,51 @@
...
 
 const PATH = `${__dirname}/sessions`;
 
-const safePath = (fn) => (token, ...args) => {
-  const callback = args[args.length - 1];
-  if (typeof token !== 'string') {
-    callback(new Error('Invalid session token'));
-    return;
-  }
+const safePath = (promise) => (token, ...args) => {
```
Comment (tshemsedinov) [inline JavaScript/storage.js:9]: `promise` is not an instance of `Promise`, it's an `AsyncFunction` so naming is confusing

File: JavaScript/storage.js [hunk 2]
```diff
@@ -1,63 +1,51 @@
...
+  async get(key) {
     const value = super.get(key);
-    if (value) {
-      callback(null, value);
-      return;
+    if (value) return value;
+    let data;
+    try {
+      data = await readSession(key);
+    } catch (err) {
```
Comment (tshemsedinov) [inline JavaScript/storage.js:27]: `err` defined but not used

**Review by tshemsedinov — CHANGES_REQUESTED** [review]
Please make separate example in new folder instead changing callback-style code. For example move `JavaScript` to `JavaScript/Callbacks` and put this example to `JavaScript/AsyncAwait`

File: JavaScript/storage.js (hunk 2 above)
Comment (OsirisAnubiz) [reply JavaScript/storage.js:27]: Would it be better to create 2 pull requests? First for fix crashing server. Second for async/await syntax.

Comment (tshemsedinov) [reply JavaScript/storage.js:27]: It would be better to open separate PRs from different branches

Comment (OsirisAnubiz) [reply JavaScript/storage.js:27]: Thank you. I will do as you said. I apologize for the inconvenience.

**Pull request closed without merging**